File: providers-ts/utils/filename-detector.ts
/**
 * Central filename type detection.
 * Mirrors FilenameDetector.kt — used by routing to classify "Unknown" entries.
 */

const VIDEO_EXTS = new Set([
  '.mkv', '.mp4', '.avi', '.webm', '.m4v', '.mov', '.wmv', '.flv', '.ts', '.m2ts',
]);
const AUDIO_EXTS = new Set([
  '.mp3', '.flac', '.aac', '.ogg', '.wav', '.wma', '.m4a', '.opus', '.alac',
]);
const IMAGE_EXTS = new Set([
  '.jpg', '.jpeg', '.png', '.webp', '.gif', '.bmp', '.tiff',
]);

export function detectType(filename: string): string {
  const lower = filename.toLowerCase();
  if (endsWithAny(lower, VIDEO_EXTS)) return 'Video';
  if (endsWithAny(lower, AUDIO_EXTS)) return 'Audio';
  if (endsWithAny(lower, IMAGE_EXTS)) return 'Image';
  return 'Unknown';
}

function endsWithAny(filename: string, exts: Set<string>): boolean {
  for (const ext of exts) {
    if (filename.endsWith(ext)) return true;
  }
  return false;
}

File: content/contract/src/main/java/com/insomnia/content/contract/FilenameDetector.kt
package com.insomnia.content.contract

/**
 * Central filename type detection.
 * Used by routing layer to classify entries with type "Unknown" from providers
 * that don't perform their own detection (e.g. SMB).
 */
object FilenameDetector {
    private val VIDEO_EXTS = setOf(
        ".mkv", ".mp4", ".avi", ".webm", ".m4v", ".mov", ".wmv", ".flv", ".ts", ".m2ts",
    )
    private val AUDIO_EXTS = setOf(
        ".mp3", ".flac", ".aac", ".ogg", ".wav", ".wma", ".m4a", ".opus", ".alac",
    )
    private val IMAGE_EXTS = setOf(
        ".jpg", ".jpeg", ".png", ".webp", ".gif", ".bmp", ".tiff",
    )

    fun detectType(filename: String): String = when {
        filename.endsWithAny(VIDEO_EXTS) -> "Video"
        filename.endsWithAny(AUDIO_EXTS) -> "Audio"
        filename.endsWithAny(IMAGE_EXTS) -> "Image"
        else -> "Unknown"
    }

    private fun String.endsWithAny(exts: Set<String>): Boolean =
        exts.any { lowercase().endsWith(it) }
}
